ramips: add support for Netgear WAX214v2

zsrv opened the pull request:
Netgear WAX214v2 is an 802.11ax (Wi-Fi 6) wireless access point.

Specifications:
* SoC: MediaTek MT7621AT
* RAM: 512 MiB
* Flash: NAND 128 MiB (ESMT PSU1GA30DT)
* Wi-Fi: 2.4/5 GHz 4T4R (MediaTek MT7915E)
* Ethernet: 1x 10/100/1000 Mbps LAN
* Switch: MediaTek MT7530 (SoC built-in)
* LEDs/Keys
  * Power (green, blue, amber)
  * LAN (green, amber)
  * WLAN 2.4GHz (green, blue)
  * WLAN 5GHz (green, blue)
  * Reset button
* USB: None
* UART: Marked J1 on board, 115200 8N1
* Power
  * 12 VDC, 1.5 A
  * IEEE 802.3at (PoE+)

Load addresses (same as Netgear WAX202):
* stock
  * 0x80010000: FIT image
  * 0x81001000: kernel image -> entry
* OpenWrt
  * 0x80010000: FIT image
  * 0x82000000: uncompressed kernel+relocate image
  * 0x80001000: relocated kernel image -> entry

MAC addresses as verified by OEM firmware:

| vendor | OpenWrt | address |
| ------ | ------- | ------- |
| eth0 | lan | label |
| ra0 | phy0 | label + 2 |
| rax0 | phy1 | label + 3 |

Installation:
* Flash the factory image by TFTP to the bootloader. NMRP can be used to TFTP without opening the case.

Revert to stock firmware:
* Flash the stock firmware to the bootloader using TFTP/NMRP.

References to WAX214v2 GPL source:
https://www.downloads.netgear.com/files/GPL/WAX214v2-V1.0.1.5-gpl-src.tar.gz

* openwrt/target/linux/ramips/dts/mt7621-ax-nand-wax214v2.dts
  DTS file for this device.

Stock boot log: [netgear_wax214v2_stock_boot_log.txt](https://github.com/openwrt/openwrt/files/13939736/netgear_wax214v2_stock_boot_log.txt)

OpenWrt boot log: [netgear_wax214v2_openwrt_boot_log.txt](https://github.com/openwrt/openwrt/files/13939740/netgear_wax214v2_openwrt_boot_log.txt)

OpenWrt boot log (commit bc312043703b17ddd6ab039ea18fea0214b6b5f0): [netgear_wax214v2_openwrt_boot_log_commit_bc31204.txt](https://github.com/openwrt/openwrt/files/13962816/netgear_wax214v2_openwrt_boot_log_commit_bc31204.txt)

OpenWrt boot log (commit 1f2553e3841c83c88e3774dadffc2b0409cbb6f0): [netgear_wax214v2_openwrt_boot_log_commit_1f2553e.txt](https://github.com/openwrt/openwrt/files/13964412/netgear_wax214v2_openwrt_boot_log_commit_1f2553e.txt)

OpenWrt boot log (commit d3977eb9d11de116d739fd5b925e40f7c3199069): [netgear_wax214v2_openwrt_boot_log_commit_d3977eb.txt](https://github.com/openwrt/openwrt/files/13964661/netgear_wax214v2_openwrt_boot_log_commit_d3977eb.txt)

OpenWrt boot log (commit 8a515d4300e423e00d7e7fc0b643bfe17e349f5f): [netgear_wax214v2_openwrt_boot_log_commit_8a515d4.txt](https://github.com/user-attachments/files/15519704/netgear_wax214v2_openwrt_boot_log_commit_8a515d4.txt)

OpenWrt boot log (commit f212d365bae7fbd75913b65f31296480d86171c0): [netgear_wax214v2_openwrt_boot_log_commit_f212d36.txt](https://github.com/user-attachments/files/16112574/netgear_wax214v2_openwrt_boot_log_commit_f212d36.txt)

zsrv commented:
Notes:
* Support for this device is based heavily on #9690
* The fix in #10366 needed to be applied for this device for uboot-envtools to work
* The WAX214v2 I have is a preproduction unit that does not have a correctly functioning LAN LED (both the green and amber LEDs are always on). According to the installation guide, the LED is supposed to be amber if the link is 10 Mbps or 100 Mbps, or green if the link is 1 Gbps. The LED on my unit turns on when the link is up, blinks when there is traffic, and turns off when the link is down. I'm not sure how this LED will behave in OpenWrt on a production unit.

neheb commented on target/linux/ramips/dts/mt7621_netgear_wax214v2.dts:
```diff
@@ -0,0 +1,211 @@
+	wifi@0,0 {
+		compatible = "mediatek,mt76";
+		reg = <0x0000 0 0 0 0>;
+		mediatek,mtd-eeprom = <&factory 0x0>;
```
should be converted to NVMEM. @DragonBluep knows more.

DragonBluep replied:
Please test if such similar change can work on your device. ref: https://github.com/openwrt/openwrt/commit/9ac891f8c44124e931c15f1257347cd8ba311a19.

zsrv replied:
I missed these changes earlier (they were also made to the Netgear WAX202). I have tried the changes made for the Netgear WAX202 in 9ac891f8c44124e931c15f1257347cd8ba311a19 and f4c33d098fdfccc741c33cfde134a2911347d17a but I see this error in the boot log with both:

```
[   13.381948] mt7915e 0000:02:00.0: WM Firmware Version: ____000000, Build Time: 20220929104145
[   13.420449] mt7915e 0000:02:00.0: WA Firmware Version: DEV_000000, Build Time: 20220929104205
[   13.539800] mt7915e: probe of 0000:02:00.0 failed with error -22
[   13.602567] kmodloader: done loading kernel modules from /etc/modules.d/*
```

My WAX214v2 EEPROM appears similar to the one in https://github.com/openwrt/openwrt/issues/13700#issuecomment-1764260302 - Byte[62] is equal to `0x07`.

DragonBluep replied:
@zsrv Please try adding `precal` part. Ref:
```
diff --git a/target/linux/ramips/dts/mt7621_netgear_eax12.dts b/target/linux/ramips/dts/mt7621_netgear_eax12.dts
index d5ad404be0b85..d0e0eb247ad94 100644
--- a/target/linux/ramips/dts/mt7621_netgear_eax12.dts
+++ b/target/linux/ramips/dts/mt7621_netgear_eax12.dts
@@ -110,6 +110,20 @@
 			label = "Factory";
 			reg = <0x100000 0x80000>;
 			read-only;
+
+			nvmem-layout {
+				compatible = "fixed-layout";
+				#address-cells = <1>;
+				#size-cells = <1>;
+
+				eeprom_factory_0: eeprom@0 {
+					reg = <0x0 0xe00>;
+				};
+
+				precal_factory_e10: precal@e10 {
+					reg = <0xe10 0x19c10>;
+				};
+			};
 		};
 
 		partition@180000 {
@@ -171,7 +185,8 @@
 	wifi@0,0 {
 		compatible = "mediatek,mt76";
 		reg = <0x0000 0 0 0 0>;
-		mediatek,mtd-eeprom = <&factory 0x0>;
+		nvmem-cells = <&eeprom_factory_0>, <&precal_factory_e10>;
+		nvmem-cell-names = "eeprom", "precal";
 	};
 };
```

zsrv replied:
@DragonBluep thank you, this worked, changes have been pushed:

```
[   12.696931] mt7915e_hif 0000:01:00.0: enabling device (0000 -> 0002)
[   12.704022] pci 0000:00:01.0: enabling device (0000 -> 0003)
[   12.709777] mt7915e 0000:02:00.0: enabling device (0000 -> 0002)
[   12.982471] mt7915e 0000:02:00.0: HW/SW Version: 0x8a108a10, Build Time: 20220929104113a
[   12.982471] 
[   13.318493] mt7915e 0000:02:00.0: WM Firmware Version: ____000000, Build Time: 20220929104145
[   13.353429] mt7915e 0000:02:00.0: WA Firmware Version: DEV_000000, Build Time: 20220929104205
[   13.475807] urngd: v1.0.2 started.
[   13.511055] mt7915e 0000:02:00.0: registering led 'mt76-phy0'
[   13.522744] mt7915e 0000:02:00.0: registering led 'mt76-phy1'
[   13.621401] kmodloader: done loading kernel modules from /etc/modules.d/*
```

DragonBluep commented on package/boot/uboot-envtools/files/ramips:
```diff
@@ -77,6 +77,7 @@ jcg,q20|\
 linksys,e7350|\
 netgear,eax12|\
 netgear,wax202|\
+netgear,wax214v2|\
```
Can you show me the output of the `hexdump -C -v -n 256 /dev/mtd1`. I saw you removed the uboot-envtools package in Makefile?

zsrv replied:
```
root@OpenWrt:/# hexdump -C -v -n 256 /dev/mtd1
00000000  97 34 66 a6 62 61 75 64  72 61 74 65 3d 31 31 35  |.4f.baudrate=115|
00000010  32 30 30 00 62 6f 6f 74  5f 6e 6d 72 70 5f 63 6c  |200.boot_nmrp_cl|
00000020  72 5f 73 74 72 5f 62 6c  6b 73 3d 6e 61 6e 64 20  |r_str_blks=nand |
00000030  65 72 61 73 65 2e 73 70  72 65 61 64 20 24 7b 73  |erase.spread ${s|
00000040  74 72 69 6e 67 74 61 62  6c 65 6f 66 66 73 65 74  |tringtableoffset|
00000050  7d 20 24 7b 73 74 72 69  6e 67 74 61 62 6c 65 66  |} ${stringtablef|
00000060  72 65 65 73 69 7a 65 7d  00 62 6f 6f 74 5f 6e 6d  |reesize}.boot_nm|
00000070  72 70 5f 77 72 5f 73 74  72 3d 6d 74 6b 75 70 67  |rp_wr_str=mtkupg|
00000080  72 61 64 65 20 73 74 72  00 62 6f 6f 74 5f 76 65  |rade str.boot_ve|
00000090  72 73 69 6f 6e 3d 32 30  32 31 30 34 30 38 00 62  |rsion=20210408.b|
000000a0  6f 6f 74 5f 77 72 5f 69  6d 67 3d 77 72 69 74 65  |oot_wr_img=write|
000000b0  69 6d 67 00 62 6f 6f 74  63 6d 64 3d 6d 74 6b 61  |img.bootcmd=mtka|
000000c0  75 74 6f 62 6f 6f 74 00  62 6f 6f 74 64 65 6c 61  |utoboot.bootdela|
000000d0  79 3d 30 00 64 65 63 72  79 70 74 5f 72 65 73 75  |y=0.decrypt_resu|
000000e0  6c 74 3d 67 6f 6f 64 00  65 74 68 61 63 74 3d 65  |lt=good.ethact=e|
000000f0  74 68 40 31 65 31 30 30  30 30 30 00 65 74 68 61  |th@1e100000.etha|
00000100
root@OpenWrt:/# 
```

This is the `fw_printenv` output:

```
root@OpenWrt:/# fw_printenv
baudrate=115200
boot_nmrp_clr_str_blks=nand erase.spread ${stringtableoffset} ${stringtablefreesize}
boot_nmrp_wr_str=mtkupgrade str
boot_version=20210408
boot_wr_img=writeimg
bootcmd=mtkautoboot
bootdelay=0
decrypt_result=good
ethact=eth@1e100000
ethaddr=94:18:65:18:E7:6F
fdtcontroladdr=9bfef730
fenv_factory=off
fenv_model=WAX214v2
fenv_region=NA
fenv_restore=yes
fileaddr=80010000
filesize=8a0000
filesize_result=good
invaild_env=no
ipaddr=192.168.1.1
mtddevname=u-boot
mtddevnum=0
mtdids=nand0=nand0
mtdparts=mtdparts=nand0:512k(u-boot),512k(u-boot-env),512k(factory),38912K(firmware),38912K(firmware_backup),13312K(CRP),4096K(Language)
netmask=255.255.255.0
partition=nand0,0
serverip=192.168.1.2
stderr=uartlite0@1e000c00
stdin=uartlite0@1e000c00
stdout=uartlite0@1e000c00
root@OpenWrt:/# 
```

The same `fw_printenv` output is seen from the stock firmware.

Without this change, this is the output:

```
root@OpenWrt:/# fw_printenv
Cannot parse config file '/etc/fw_env.config': No such file or directory
Failed to find NVMEM device
root@OpenWrt:/# 
```

#10366 has information about why this change was needed for the Netgear WAX202. The contributor of that change suggested that it is probably safer to leave the Config partition read-only, and I figured that they removed the uboot-envtools package for the WAX202 since changes would not be able to be made. I followed suit when adding support for the WAX214v2; would a different approach be better?

DragonBluep replied:
Hi! Please use u-boot-env `ethaddr` MAC address as the base MAC instead of the `mac`. In this way, we can move MAC settings to dts in the future.

zsrv replied:
Okay, done

hauke replied:
Do you want to switch to `NVMEM_U_BOOT_ENV` already now, or should we do it later for all 3 devices?

zsrv replied:
I don't mind either way (but I would need help if we should do it for this device now), but I found https://github.com/openwrt/openwrt/pull/14182 - should we wait for this to be finished first?

neheb replied:
+1 for u-boot,env. That PR is stalled. Might as well do it here.

zsrv replied:
Okay, I've made an attempt at switching to `NVMEM_U_BOOT_ENV` and `compatible = "u-boot,env";` in the latest push, and have attached the latest OpenWrt boot log in the PR description (for commit 8a515d4300e423e00d7e7fc0b643bfe17e349f5f).

One problem I encountered is that removing [this line](https://github.com/openwrt/openwrt/commit/8a515d4300e423e00d7e7fc0b643bfe17e349f5f#diff-a4a13db5c96026e0a9d0909fa19257b892affe1f5944275a7f4ce851eb90ca79R82) made `fw_setenv` fail with the following error:

```
root@OpenWrt:/# fw_setenv bootdelay 5
Can't open /sys/bus/nvmem/devices/u-boot-env0/nvmem: Permission denied
Error: can't write fw_env to flash
```

Leaving this line in made `fw_setenv` work properly again (a test of this is in the latest OpenWrt boot log). I'm not sure if there is something I was missing.

neheb replied:
So mac address is correct without the line?

zsrv replied:
I rebuilt without the line again as a test and the mac address is correct without it.

neheb replied:
Weird. Does fw_printenv work without it too?

zsrv replied:
Yes, fw_printenv works with or without the line.

neheb replied:
I would guess envtools needs a patch to deal with this then. ping @Ansuel @rmilecki

neheb replied:
Sorry to do this, but please revert. I've confirmed the nvmem uboot env driver does not support writing. This will have to wait for some other time.

In other news, I restarted the other PR.

zsrv replied:
Okay, I've reverted the nvmem changes. A new OpenWrt boot log has been added to the description (for commit f212d365bae7fbd75913b65f31296480d86171c0).

DragonBluep commented on target/linux/ramips/dts/mt7621_netgear_wax214v2.dts:
```diff
@@ -0,0 +1,226 @@
+		compatible = "gpio-leds";
+
+		led_power_green: power_green {
+			label = "green:power";
```
Please use `function` && `color` properties instead of the `label`. examples:
https://git.openwrt.org/?p=openwrt%2Fopenwrt.git&a=search&h=HEAD&st=grep&s=function+%3D+LED

zsrv replied:
Done, except for LEDs green:wifia and green:wifin for which I could not find suitable LED function macros.

raenye replied:
`function = LED_FUNCTION_WLAN` and use `function-enumerator` to differentiate.

zsrv replied:
Done - it appears that functions LED_FUNCTION_WLAN_2GHZ and LED_FUNCTION_WLAN_5GHZ have since been added, so I used those. LED entries appear like this now:

```
root@netgear-wax214v2:~# ls -1 /sys/class/leds/
amber:power
blue:power
green:power
green:wlan-2ghz
green:wlan-5ghz
mt76-phy0
mt76-phy1
```

DragonBluep commented on target/linux/ramips/dts/mt7621_netgear_wax214v2.dts:
```diff
@@ -0,0 +1,226 @@
+			read-only;
+		};
+
+		factory: partition@100000 {
```
`factory:` label can be removed.

zsrv replied:
Done

DragonBluep commented on target/linux/ramips/dts/mt7621_netgear_wax214v2.dts:
```diff
@@ -0,0 +1,226 @@
+
+&gmac1 {
+	status = "okay";
+	label = "lan";
```
I'm not sure if it will cause some potential problems. If the device only has one LAN port, I suggest using a single port switch on gmac0. @arinc9 is an expert for this. Perhaps we can wait for his comments.

arinc9 replied:
This actually will work. The DSA driver must run for this to work, and that will still happen without any user ports being described on the device tree. Though I wouldn't do this for a device with only one user port. I would suggest to get rid of &gmac1 and &mdio here, and describe the port on the switch node instead.

```
&switch0 {
	ports {
		port@0 {
			status = "okay";
			label = "lan";
		};
};
```

zsrv replied:
Okay, done as suggested

DragonBluep commented on target/linux/ramips/mt7621/base-files/etc/board.d/01_leds:
```diff
@@ -183,6 +183,10 @@ netgear,r7450)
 	;;
+netgear,wax214v2)
+	ucidef_set_led_wlan "wifia_blue" "WiFi 2.4GHz (Blue)" "mt76-phy0" "phy0tpt"
+	ucidef_set_led_wlan "wifin_blue" "WiFi 5GHz (Blue)" "mt76-phy1" "phy1tpt"
```
The default mt76 LED trigger is tpt so I think these configs can be removed.

zsrv replied:
You are right, I have removed these.

zsrv commented on target/linux/ramips/dts/mt7621_netgear_wax214v2.dts:
```diff
@@ -0,0 +1,227 @@
+			gpios = <&gpio 13 GPIO_ACTIVE_LOW>;
+		};
+
+		led_wifia_green: wifia_green {
```
Done

zsrv commented on target/linux/ramips/dts/mt7621_netgear_wax214v2.dts:
```diff
@@ -0,0 +1,227 @@
+
+&state_default {
+	gpio {
+		groups = "uart3", "uart2", "jtag", "wdt";
```
Done

hauke commented on target/linux/ramips/image/mt7621.mk:
```diff
@@ -1990,6 +1990,24 @@ define Device/netgear_wax202
+  $(Device/nand)
+  DEVICE_VENDOR := NETGEAR
+  DEVICE_MODEL := WAX214v2
+  DEVICE_PACKAGES := kmod-mt7915-firmware -uboot-envtools
```
Please do not remove `uboot-envtools`. You defined the u-boot settings.

zsrv replied:
Okay, should I also remove `read-only` from the Config partition?

hauke replied:
if setting uboot environment variables works, make it writable too please.

zsrv replied:
Okay, done

neheb commented on target/linux/ramips/dts/mt7621_netgear_wax214v2.dts:
```diff
@@ -0,0 +1,228 @@
+		};
+
+		partition@80000 {
+			label = "Config";
```
Why is this not read-only? I don't see counter manipulations or anything.

zsrv replied:
Originally, I followed what was done for the Netgear WAX202 and made the Config partition read-only and removed the uboot-envtools package (if I recall, the rationale for that was that the uboot environment variables should not be manipulated). It was suggested in [this discussion](https://github.com/openwrt/openwrt/pull/14401#discussion_r1460582412) that if the uboot environment variables could be set, then the partition storing them should not be read-only and uboot-envtools should not be removed.

neheb replied:
hmm OK.

zsrv commented:
I have added support for reading/writing to the second u-boot environment that is present in the Config partition. Read and write operations seem to work fine on it, see the latest OpenWrt boot log in the PR description (for commit f212d365bae7fbd75913b65f31296480d86171c0).

neheb commented:
Now I'm curious if the WAX202 has such a thing...

neheb commented:
```
Author name (zsrv) need to be your real name 'firstname lastname'
Commit subject line seems ok (ramips: add support for Netgear WAX214v2)
Signed-off-by match author
```

zsrv committed:
ramips: add support for Netgear WAX214v2

Netgear WAX214v2 is an 802.11ax (Wi-Fi 6) wireless access point.

Specifications:
* SoC: MediaTek MT7621AT
* RAM: 512 MiB
* Flash: NAND 128 MiB (ESMT PSU1GA30DT)
* Wi-Fi: 2.4/5 GHz 4T4R (MediaTek MT7915E)
* Ethernet: 1x 10/100/1000 Mbps LAN
* Switch: MediaTek MT7530 (SoC built-in)
* LEDs/Keys
  * Power (green, blue, amber)
  * LAN (green, amber)
  * WLAN 2.4GHz (green, blue)
  * WLAN 5GHz (green, blue)
  * Reset button
* USB: None
* UART: Marked J1 on board, 115200 8N1
* Power
  * 12 VDC, 1.5 A
  * IEEE 802.3at (PoE+)

Load addresses (same as Netgear WAX202):
* stock
  * 0x80010000: FIT image
  * 0x81001000: kernel image -> entry
* OpenWrt
  * 0x80010000: FIT image
  * 0x82000000: uncompressed kernel+relocate image
  * 0x80001000: relocated kernel image -> entry

MAC addresses as verified by OEM firmware:

vendor   OpenWrt   address
eth0     lan       label
ra0      phy0      label + 2
rax0     phy1      label + 3

Installation:
* Flash the factory image by TFTP to the bootloader.
  NMRP can be used to TFTP without opening the case.

Revert to stock firmware:
* Flash the stock firmware to the bootloader using TFTP/NMRP.

References to WAX214v2 GPL source:
https://www.downloads.netgear.com/files/GPL/WAX214v2-V1.0.1.5-gpl-src.tar.gz

* openwrt/target/linux/ramips/dts/mt7621-ax-nand-wax214v2.dts
  DTS file for this device.

Signed-off-by: Borys Zhukov <borys@zhukov.org>
Link: https://github.com/openwrt/openwrt/pull/14401
Signed-off-by: Hauke Mehrtens <hauke@hauke-m.de>

zsrv commented:
> ```
> Author name (zsrv) need to be your real name 'firstname lastname'
> Commit subject line seems ok (ramips: add support for Netgear WAX214v2)
> Signed-off-by match author
> ```

Okay, done
